Lesson3

Robilko opened the pull request:
1. Все кнопки обработаны
2. Создан отдельный класс где вынесена вся логика работы калькулятора и отделена от основной активити. Этот класс имплементирует Сериализацию, соответствено при смене ориентации восстанавливается оттуда.
3*. Макет для ландшафтной ориентации создан.

Все замеченные баги в логике калькулятора устранены в процессе. 

Так как за основу взят базовый калькулятор, без замаха на инженерный, то все операции производятся только в одно действие. 
Единственный нюанс в логике, который хотел бы изменить в будущем , это чтобы после ввода второго аргумента и нажатии после этого  кнопки (+, -, * или /) вместо = , он автоматом считал выражение и уже к нему продолжал выполнять последующий действия
![1622138786911](https://user-images.githubusercontent.com/75025794/119876825-f71adb80-bf30-11eb-97ae-723574bc6f19.jpg)
![1622138786907](https://user-images.githubusercontent.com/75025794/119876823-f71adb80-bf30-11eb-9589-73cd3759b927.jpg)

Robilko committed:
Выполнено задание к уроку 2

Robilko committed:
Созданы иконки к приложению

Robilko committed:
Созданы иконки к приложению

Robilko committed:
Урок 3. Написана основная логика работы приложения и клавиш. Нужно исправить пару ошибок и дописать функционал нескольких клавиш

Robilko committed:
Урок 3.
+ Исправлены ошибки в логике + дописана логика для остальных кнопок.
+ Создал несколько анонимных классов для кнопок и для действий, чтобы в классе калькулятора не было обращений к ресурсам напрямую.

Robilko committed:
Урок 3.
+добавлена ландшафтная ориентация экрана

Robilko committed:
Урок 3.
+немного подправлен дизайн

Robilko committed:
Урок 3.
+устранен баг - после полной очистки поля при нажатии кнопки с действием программу выкидывало

Robilko committed:
Урок 3.
+поправил код

Robilko committed:
Урок 3 + исправил все найденные баги в логике работы калькулятора. Прилизал вывод результата (убрал нули дробей, если число целое)

Robilko committed:
Урок 3

Robilko committed:
Урок 3 переделал метод onActionPressed через switch вместо if

Robilko committed:
Урок 3 +добавил сохранение состояния при смене ориентации экрана.

EvgeniyChufarnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,218 @@
+public class CalculatorModel implements Serializable {
+    private double firstArg;
+    private boolean hasDot = false;
+    private final int ARG_MAX_LENGTH = 20;
```
Если набрать оба числа по 20 символов, то второе не полностью входит в истории

EvgeniyChufarnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,218 @@
+    private String doublePrimeToInt(double number) {
+        String result = "";
+        if (number % 1 == 0)
+            result +=(int)number;
```
при касте из double в int может измениться число, например если набрать 1-999999999999, то выдаст -2147483648, т.к. число не поместилось в int

EvgeniyChufarnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,218 @@
+                }
+                break;
+
+            case  ActionButton.PERCENT:
```
после действия процент можно поставить ещё 1 точку, и дальше если нажать ещё действие, приложение вылетает

EvgeniyChufarnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,218 @@
+                            break;
+                        case ActionButton.DIVISION:
+                            if (firstArg == 0 || secondArg == 0) {
+                                ARGUMENT.append(0);
```
таким образом получается, что при делении на 0 результат тоже 0, что не очень верно математически

Robilko replied:
Спасибо, исправил!

Robilko committed:
Урок 3 +исправлено два бага: 1. деление на ноль отображается предупреждающий текст. 2. Исправлен баг, когда при действии % можно было добавить еще точку и приложение вылетало

Robilko committed:
Урок 3 +исправлен баг - была возможность ставить несколько точек после выбора матем действий, после чего программа вылетала.
+ исправлен баг с отбрасыванием дробной части, если число целое. Использован long вместо int

Robilko committed:
Урок 3 Рефакторинг кода

Robilko committed:
Урок 3

kshalnov commented on app/build.gradle:
```diff
@@ -8,7 +8,7 @@ android {
     defaultConfig {
         applicationId "com.robivan.calculator"
-        minSdkVersion 23
+        minSdkVersion 26
```
а почему?

kshalnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,198 @@
+
+public class CalculatorModel implements Serializable {
+    private double firstArg, secondArg;
+    String DIV_BY_ZERO = "Разделить на ноль нельзя";
```
В идеале вынести в ресурсы

kshalnov replied:
И почему не private?

Robilko replied:
вынес

kshalnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,198 @@
+    private double firstArg, secondArg;
+    String DIV_BY_ZERO = "Разделить на ноль нельзя";
+
+    private final StringBuilder ARGUMENT = new StringBuilder();
```
большими буквами именуют константы - неизменяемые статичные поля. Тут формально ссылка final, но содержимое при этом меняется

kshalnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,198 @@
+
+    private int actionSelected;
+
+    private enum State {
```
откуда вы все этот State берёте? уже в третий раз точно вижу)

kshalnov commented on app/src/main/java/com/robivan/calculator/Util.java:
```diff
@@ -0,0 +1,44 @@
+package com.robivan.calculator;
+
+public abstract class Util {
```
таким образом на конкретные id и на ui сейчас завязано три класса: активити, утил и калькулятор. В идеале надо скрыть особенности разметки внутри активити

kshalnov commented on app/src/main/java/com/robivan/calculator/MainActivity.java:
```diff
@@ -1,14 +1,70 @@
 public class MainActivity extends AppCompatActivity {
+    private CalculatorModel calculator;
+    private TextView display, resultScreen;
+    protected final String CALC_KEY = MainActivity.class.getCanonicalName() + "calc_key";
```
private static

kshalnov commented on app/src/main/java/com/robivan/calculator/CalculatorModel.java:
```diff
@@ -0,0 +1,198 @@
+
+    }
+
+    public void onActionPressed(int actionId) {
```
очень большой метод со всей логикой, в идеале его бы разбить на несколько.
